gguf-py/gguf/gguf_writer.py

```diff
@@ -106,36 +141,72 @@ def open_output_file(self, path: os.PathLike[str] | str | None = None) -> None:
             self.path = path
 
         if self.path is not None:
-            if self.fout is not None:
-                self.fout.close()
-            self.fout = open(self.path, "wb")
+            self.fout = []
+            for fout in self.format_shard_names(self.path):
+                self.fout.append(open(fout, "wb"))
             self.state = WriterState.EMPTY
 
+    def print_plan(self, path: os.PathLike[str] | str | None = None) -> None:
+        logger.info("Writing the following files:")
+        filenames = self.format_shard_names(path)
+        assert len(filenames) == len(self.tensors)
+        for name, tensors in zip(filenames, self.tensors):
+            logger.info(f"{name}: n_tensors = {len(tensors)}, total_size = {GGUFWriter.format_n_bytes_to_str(sum(ti.nbytes for ti in tensors.values()))}")
+
+        if self.dry_run:
+            logger.info("Dry run, not writing files")
+            exit()
+
+    def add_shard_kv_data(self) -> None:
+        if len(self.tensors) == 1:
+            return
+
+        total_tensors = sum(len(t) for t in self.tensors)
+        for i in range(len(self.fout)):
+            # just see whether it exists
+            try:
+                self.kv_data[i]
+            except IndexError:
+                self.kv_data.append(dict())
+            self.kv_data[i][Keys.Split.LLM_KV_SPLIT_NO] = GGUFValue(i, GGUFValueType.UINT16)
+            self.kv_data[i][Keys.Split.LLM_KV_SPLIT_COUNT] = GGUFValue(len(self.fout), GGUFValueType.UINT16)
+            self.kv_data[i][Keys.Split.LLM_KV_SPLIT_TENSORS_COUNT] = GGUFValue(total_tensors, GGUFValueType.INT32)
+
     def write_header_to_file(self, path: os.PathLike[str] | str | None = None) -> None:
+        self.verify_arguments()
+        self.print_plan(path)
```

Comment:
`self.path` really should be `Path | None`: `fname_out` is a `Path` in `convert-hf-to-gguf.py` so there's no reason for it to be anything else (other than `None`) in `GGUFWriter`.

Reply:
This also makes `path` pointless in the signature of `open_output_file` (unless we want to duplicate the code). Is removing it acceptable?

Reply:
> `self.path` really should be `Path | None`

True. I agree for `self.path`. But I think the `path` argument of `GGUFWriter.__init__` should stay a `Pathlike[str] | str | None`, because some scripts like `gguf-py/examples/writer.py` and `examples/llava/convert-image-encoder-to-gguf.py` use a `str` for the path.

https://github.com/ggerganov/llama.cpp/blob/e95beeb1fc4621826ddd616776dbdf717366bf5c/examples/llava/convert-image-encoder-to-gguf.py#L167-L168

https://github.com/ggerganov/llama.cpp/blob/e95beeb1fc4621826ddd616776dbdf717366bf5c/gguf-py/examples/writer.py#L16